src/rmp/src/Restructure.cpp

```diff
@@ -276,21 +284,44 @@ void Restructure::runABC()
     output_blif_file_name_
         = work_dir_name_ + std::string(block_->getConstName())
           + std::to_string(curr_mode_idx) + "_crit_path_out.blif";
+    std::string abc_log_name = logfile_ + std::to_string(curr_mode_idx);
 
+    int level_gain = 0;
+    float delay = std::numeric_limits<float>::max();
     int num_instances = 0;
-    bool status
-        = blif_.inspectBlif(output_blif_file_name_.c_str(), num_instances);
-    logger_->report("Optimized to {} instances in iteration {}.",
-                    num_instances,
-                    curr_mode_idx);
-    if (status && num_instances < best_inst_count) {
-      best_inst_count = num_instances;
-      best_blif = output_blif_file_name_;
+    bool status = readAbcLog(abc_log_name, level_gain, delay);
+    if (status) {
+      status
+          = blif_.inspectBlif(output_blif_file_name_.c_str(), num_instances);
+      logger_->report("Optimized to {} instances in iteration {} with max path depth decrease of {}, delay of {}.",
+                      num_instances,
+                      curr_mode_idx, level_gain, delay);
+    }
+    if (status) {
+      if (modes[0] < Mode::DELAY_1) {
+        if (num_instances < best_inst_count) {
+          best_inst_count = num_instances;
+          best_blif = output_blif_file_name_;
+        }
+      } else {
+        if (level_based && level_gain > best_level_gain) {
+          best_level_gain = level_gain;
+          best_blif = output_blif_file_name_;
+        }
+        // Using only DELAY_4 for delay based gain
+        if ((curr_mode_idx == modes.size()-1) || (!level_based && delay < best_delay_gain)) {
```

Comment:
modes.size()-1 may not always be DELAY_4.  Explain why only this one mode is used.